guides/plugins/plugins/checkout/cart/change-price-of-item.md

```diff
@@ -0,0 +1,306 @@
+
+        foreach ($products as $product) {
+            $key = $this->buildKey($product->getReferencedId());
+
+            // no overwritten price? continue with next product
+            if (!$data->has($key) || $data->get($key) === null) {
+                continue;
+            }
+
+            $newPrice = $data->get($key);
+
+            // build new price definition
+            $definition = new QuantityPriceDefinition(
+                $newPrice,
+                $product->getPrice()->getTaxRules(),
+                $context->getCurrency()->getDecimalPrecision(),
+                $product->getPrice()->getQuantity(),
+                true
+            );
+
+            // build CalculatedPrice over calculator class for overwritten price
+            $calculated = $this->calculator->calculate($definition, $context);
+
+            // set new price into line item
+            $product->setPrice($calculated);
+            $product->setPriceDefinition($definition);
+        }
+    }
+
+    private function filterAlreadyFetchedPrices(array $productIds, CartDataCollection $data): array
+    {
+        $filtered = [];
+
+        foreach ($productIds as $id) {
+            $key = $this->buildKey($id);
+
+            // already fetched from database?
+            if ($data->has($key)) {
+                continue;
+            }
+
+            $filtered[] = $id;
+        }
+
+        return $filtered;
+    }
+
+    private function buildKey(string $id): string
+    {
+        return 'price-overwrite-'.$id;
+    }
+}
+
+```
+{% endcode %}
+
+First of all, note the second interface we implemented, next to the `CartDataCollectorInterface`.
+We also added a constructor in order to save the `QuantityPriceCalculator`.
+
+But now, let's have a look at the `process` method.
+You should already be familiar with most of its parameters, since they're mostly the same with those of the collector. 
+Yet, there's one main difference: Next to the `$original` `Cart`, you've got another `Cart` parameter being called `$toCalculate`here.
+Make sure to do all the changes on the `$toCalculate` instance, since this is the cart that's going to be considered in the end.
+The `$original` one is just there, because it may contain necessary data for the actual cart instance.
+
+Now let's have a look inside the `process` method.
+
+We start by filtering all line items down to only products, just like we did in the `collect` method.
+Then we're iterating over all products found, building the unique key, which is necessary for fetching the new price from the `CartDataCollector`.
+
+If there's no price to be processed saved in the `CartDataCollector`, there's nothing to do here.
+Otherwise, we're fetching the new price, we're building a new instance of a `QuantityPriceDefinition` containing the new price.
+Using that instance, we can calculate the actual new price using the previously injected `QuantityPriceCalculator`.
+
```

Comment:
A hint could be useful here, stating that you should'nt do any external or database queries in the process method.
Even more, beacuse this is specifically asked in the trainings.

'Do not query the DB here. Use the collect method of the collector instead...'